[Mobile Payments] Consolidate CardPresent and PaymentGateway Stores; Switch Between WCPay and Stripe Backends Within

allendav opened the pull request:
Closes: #5853
Closes: #5534
Closes: #5536
Closes: #5531

### Description
- Combines PaymentGatewayAccountStore and CardPresentPaymentStore, moving the actions in the former into the latter
- Consolidates their unit tests as well
- Changes onboarding to defer loading accounts until after it knows which plugin is active
- Also tells the CardPresentPaymentStore which backend to use for things like fetching accounts, connecting readers
- Defaults to WCPay backend

### Unfinished Business
- Define a protocol for the two remotes to simplify the code in CardPresentPaymentStore
- Add unit tests to cover the logic added to CardPresentPaymentStore and CardPresentPaymentsOnboardingUseCase - see also #5866 and #5867
- **I'm going to move these into subsequent PR(s) - this one is already so big.**

### Please note
- This does NOT hook up these changes to the Order flow. Capturing payments through Stripe powered stores will come in a later PR.

### Testing instructions
- First, ensure you can connect to a reader on a WCPay powered store
- Next, set up a store (if you haven't already) using the Stripe extension.
- IMPORTANT: Modify woocommerce-gateway-stripe/includes/admin/class-wc-rest-stripe-account-controller.php so that get_account_summary also returns keys of is_live (false) and test_mode (true) if necessary - see https://github.com/woocommerce/woocommerce-gateway-stripe/issues/2262 for more details
- Enable the Stripe IPP feature under Settings > Experimental
- Then, ensure you can connect to a reader on that Stripe powered store

### Screenshots
N/A

---
- [ ] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

allendav committed:
Combine PaymentGatewayAccountStore and CardPresentPaymentStore in prep for gateway switching

ctarda commented:
It is looking pretty good. 

I figure you haven't reached this point yet, but just to make sure we don't miss it:
* when calling startDiscovery in the reader service, we pass a token provider that wraps the remote. That token provider will need to be different for Stripe and WCPay
* when calling reset into the service, we would need to do something, perhaps point the store to a default remote?

allendav committed:
Add actions to switch CPP Store between WCPay and Stripe backends

allendav committed:
Direct account, customer and capture actions to appropriate backend

allendav committed:
Add a Card Reader Config Provider for Stripe backend

allendav committed:
Update onboarding to set active backend for CPP store based on active plugin

allendav committed:
Only attempt Stripe if the experiment is enabled; debugging prints

allendav committed:
Correct location endpoint path in stripe remote

allendav committed:
Remove prints added for debugging

allendav committed:
Remove remaining bits of PaymentGatewayAccountStore

allendav committed:
Correct url suffix in stripe remote unit tests

ctarda commented on WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person Payments/CardPresentPaymentsOnboardingUseCase.swift:
```diff
@@ -104,24 +144,24 @@ private extension CardPresentPaymentsOnboardingUseCase {
-        }
-        group.enter()
-        stores.dispatch(paymentGatewayAccountsAction)
+//        // We need to sync payment gateway accounts to see if WCPay is set up correctly
```
This could be removed

allendav replied:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/f2f6a89eeaf4b89edace55186e922f83aacbd46a

ctarda commented on Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:
```diff
@@ -4,6 +4,31 @@
+
+    /// Get a Stripe Customer for an order.
+    ///
+    case fetchOrderCustomer(siteID: Int64, orderID: Int64, onCompletion: (Result<WCPayCustomer, Error>) -> Void)
```
We might want to revisit the naming of WCPayCustomer in the future?

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5903

ctarda commented on Yosemite/Yosemite/Actions/CardPresentPaymentAction.swift:
```diff
@@ -4,6 +4,31 @@
 public enum CardPresentPaymentAction: Action {
+    /// Switches the store to use WCPay as the backend. This is also the default.
+    ///
+    case useWCPay
```
I don't mind declaring these two cases. It makes the code extremely declarative. 

One potential way to make this a bit more abstract could be to declare an enumeration (i.e. PaymentGateway) with two cases, and rename these two actions into a single action: `useGateway(gateway: PaymentGateway)

It could kind of future proof the API for a potential third gateway. On the other hand, a third gateway does not sound like something we would add in the nearly future, so 🤷🏼

allendav replied:
Good idea, but let's cross that bridge if/when we get to a third gateway? :)

ctarda replied:
Right. There is one more argument supporting doing this soon, without waiting for a third plugin: serializing an enumeration to user defaults is simple. Have in mind we are going to have to save the plugin selected by the user to user defaults, and materialize it again from user defaults before performing any other operations. We could just save and read a boolean flag, but abstracting that behind an enumeration is very low effort for a significant benefit, imo.

So, sure, let's defer it, but not for too long.

ctarda commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -73,9 +97,28 @@ public final class CardPresentPaymentStore: Store {
     func startCardReaderDiscovery(siteID: Int64, onReaderDiscovered: @escaping (_ readers: [CardReader]) -> Void, onError: @escaping (Error) -> Void) {
         do {
-            try cardReaderService.start(WCPayTokenProvider(siteID: siteID, remote: self.remote))
+            switch usingBackend {
```
You mentioned in the description making both remotes implement a common interface in a future PR.

I would take that one step further, and I would suggest extracting the logic to decide which backend we are calling into to a separate entity. For example, that could be a class with a single public method similar to this (naming is terrible, but just to illustrate the idea):
```
func getRemote(gateway: PaymentGateway) ->  PaymentsRemote {
switch gateway {
case .stripe: return stripeRemote
case .wcpay: return wcpayRemote
}
}
```

This class could be declared within this store, so it is completely private.

allendav replied:
Opened https://github.com/woocommerce/woocommerce-ios/issues/5904

ctarda commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -73,9 +97,28 @@ public final class CardPresentPaymentStore: Store {
-            try cardReaderService.start(WCPayTokenProvider(siteID: siteID, remote: self.remote))
+            switch usingBackend {
+            case .wcpay:
+                print("==== starting card reader discovery using a wcpay token provider")
```
It would be good to remove these prints, if not now, in a follow-up PR

allendav replied:
Oops. Meant to remove all of them.

allendav replied:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/d8db38f438a71f5bc36e6b439dc882a2338ca6c1

ctarda commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -291,3 +385,224 @@ private extension CardReaderConfigError {
+            return
+        }
+
+        reset() // Concern: asynchronicity
```
I am not sure I would call reset here. If we do, the API docs of these methods should, in my opinion, clarify that there is going to be an implicit call to reset.

I think it would be slightly preferable to leave the responsibility of resetting the SDK to an upper layer.

Also, the Terminal SDK tends to error quite often for attempting operations when it's busy, so we should make sure we do not introduce instability

allendav replied:
I've removed the resets from the service layer in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/fc0132753e9c15b45801a177b871b6d904fdc9a1 but I haven't added them to the upper layer yet. During testing I noticed that even when switching between WCPay powered and Stripe powered stores, the Terminal SDK was continuing to ask the previous service for tokens and location.

I'll continue digging into this. I think we might have missed something in https://github.com/woocommerce/woocommerce-ios/pull/4147

ctarda replied:
Could it because there is no call to reset() before switching the store?

Looking at the implementation of setConfigProvider in StripeCardReaderService, there is this:

```
        if !Terminal.hasTokenProvider() {
            Terminal.setTokenProvider(tokenProvider)
        }
```

ctarda commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -291,3 +385,224 @@ private extension CardReaderConfigError {
+                self.deleteStaleAccount(siteID: siteID, gatewayID: WCPayAccount.gatewayID)
+            }
+
+            onCompletion(.success(())) // TODO - consider bringing back Error, although callers don't currently use it
```
Please don't make this a TODO, make it a ticket. Even if clients don't use it, it is good to surface as many errors as possible

allendav replied:
I just went ahead and brought back failure - https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/5c19233add147d6e8ac7b59c98533e66999fe17e

ctarda commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -291,3 +385,224 @@ private extension CardReaderConfigError {
+                self.deleteStaleAccount(siteID: siteID, gatewayID: StripeAccount.gatewayID)
+            }
+
+            onCompletion(.success(())) // TODO - consider bringing back Error, although callers don't currently use it
```
Same comment about making this a ticket instead of a TODO

allendav replied:
I just went ahead and brought back failure - https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/5c19233add147d6e8ac7b59c98533e66999fe17e

ctarda commented on Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift:
```diff
@@ -73,9 +97,28 @@ public final class CardPresentPaymentStore: Store {
+            switch usingBackend {
+            case .wcpay:
+                print("==== starting card reader discovery using a wcpay token provider")
+                try cardReaderService.start(WCPayTokenProvider(siteID: siteID, remote: self.remote))
```
I would feel more comfortable merging this into trunk if we checked the feature flag in all these methods, and short-circuited defaulting to using the wcpayRemote when the feature flag is not active.

allendav replied:
Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5868/commits/58187d63416200c480bdb8ab75e71be097adcbc1

ctarda reviewed (approved):
Thanks for tackling this @allendav . I appreciate the iterative approach and how you have tried to find a reasonable solution given the trade-offs.

I left some comments for you to consider. The only thing I will suggest doing before merging is short-circuiting all the operations, defaulting to use WCPayRemote, when the feature flag is not enabled.

In that sense, extracting the logic to select the remote that should be used to a different class before merging, as I mentioned in a comment, might be helpful. But I'll leave that to you.

In any case, if you agree, I think it would be good to at least take the feature flag into consideration: safety first!

:shipit:

allendav committed:
Return failure to load account to completion

allendav committed:
Remove debug prints accidentally left in

allendav committed:
Remove commented out block accidentally left in

allendav committed:
Pass stripeExtensionInPersonPayments feature flag into CardPresentPaymentStore

allendav committed:
Remove resets from service layer

allendav commented:
@ctarda - thank you for your review - I've addressed all the feedback but I am now going to see if this is a problem apart from this PR: https://github.com/woocommerce/woocommerce-ios/pull/5868#discussion_r785181633 (which I suspect it is)

If that turns out to be the case, we would be able to move forward with this PR and address that bug separately

ctarda commented:
@allendav  can we please merge this as it is, regardless of [this ](https://github.com/woocommerce/woocommerce-ios/pull/5868#discussion_r785181633) and my reply to it? 

This PR is kind becoming a blocker for making further progress

allendav commented:
Merged!
